---
title: "Filtering data with vcfR"
author: "Brian J. Knaus"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Filtering data with vcfR}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---


Most variant calling pipelines result in files containing variant information.
The [vcf format](http//www.1000genomes.org/wiki/analysis/variant%20call%20format/vcf-variant-call-format-version-41 "vcf format at 1k genomes") is an increasingly popular format for this data.
These files are typically intended to be filtered as an attempt to remove false positives, or otherwise problematic sites.
The R package vcfR provides tools to facilitate this filtering as well as to visualize the effects of choices made during this process.


## Data input


The input of data was covered in the vignette 'Introduction to vcfR.
If you haven't read this vignette, you can with:

```{r, eval=FALSE}
vignette("intro_to_vcfR", package="vcfR")
```

.

For this vignette, we'll use the example data provided with vcfR.


```{r}
library(vcfR)
data(vcfR_example)
pinf_mt <- create_chrom(name='pinf_mt', vcf=pinf_vcf, seq=pinf_dna, ann=pinf_gff)
pinf_mt <- proc_chrom(pinf_mt, verbose=FALSE)
```


We can now visualize this data.


```{r, fig.height=7, fig.width=7}
chromoqc(pinf_mt)
```



## Getting to know your vcf data


A vcf file can be thought of as having three regions: a meta region, a fixed region and a genotype region.
The meta region is at the top of the file.
Each line of the meta region begins with a double pound sign ('##').
The information in the meta region defines the abbreviations used elsewhere in the file.
Below the meta region, the data are tabular.



The first eight columns contain the fixed region and are titled CHROM, POS, ID, REF, ALT, QUAL, FILTER and INFO.
This is per variant information which is 'fixed', or the same, over all samples.


```{r}
head(pinf_vcf)
```




```{r}
head(pinf_mt)
```



