---
title: "Visualizing vcf data II"
author: "Brian Knaus"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Visualization II}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---


In the vignette 'Visualizing vcf data I' we began to explore how to plot information contained in vcf format data.
Here we build on this by exploring data based on each sample's genotype information.


## Data


As in other vignettes, we begin by loading the example data.



```{r}
library(vcfR)
data(vcfR_example)
pinf_mt <- create_chrom(name='pinf_mt', vcf=pinf_vcf, seq=pinf_dna, ann=pinf_gff, verbose=FALSE)
pinf_mt <- proc_chrom(pinf_mt, verbose=FALSE, win.size=1e3)
```



## Genotype data

In the vignette Visualizing vcf data I we explored summaries of variants made across all samples.
If all samples are fairly homogenous in the qualities of these metrics, than a summary across all samples may be appropriate.
However, if a few samples have very low quality, or some samples are exceptional, this may be masked by looking at summaries across all samples.
By looking at the individual samples, these differences may be discovered.



In order to look at the per sample information, we first need to determine what sort of data we have there.
We can use the function head() to summarize the information in our Chrom object.


```{r}
head(pinf_mt)
```



Recall from the vignette 'vcf data' that the genotype information begins at column nine.
Column nine specifies the format of all subsequent columns.
All columns subsequent to column nine contain information for each sample where each column corresponds to a sample.

We see form the output of head() that our genotype data contains GT:PL:GQ data.
This is the genotype, phred-scaled likelihoods for each genotype and genotype quality.
Observation of the first few rows of genotype information shows the format as 1/1:254,255,0:99.
The genotypes are in the form of 0/0, 0/1 and 1/1 for homozygote for the reference allele, heterozygote and homozygote for the alternate allele.
Because there is a forward slash we won't be able to directly treat this as numeric information.
The phred-scaled likelihoods appear as comma delimited integers.
Again, because there is a non-numeric character (the commas) we won't be able to treat these directly as numeric data.
The genotype quality appears as integers which we can treat as numeric data, so we'll use that as our first example.


## Using extract.gt



```{r}
gq <- extract.gt(pinf_mt, element="GQ", as.numeric=TRUE)
head(gq)
```




```{r}
heatmap.bp(gq)
```




```{r}
gqw <- windowize_NM(gq, pos=pinf_mt@var.info$POS, starts=pinf_mt@win.info$start, ends=pinf_mt@win.info$end)
heatmap.bp(gqw)
```





